text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+
+  createHelper(definition: HelperDefinition, args: TemplateArgs): HelperStateBucket;
+
+  getValue?(bucket: HelperStateBucket, args: TemplateArgs): unknown;
+
+  runEffect?(bucket: HelperStateBucket, args: TemplateArgs): void;
+
+  getDestroyable?(bucket: HelperStateBucket): object;
+}
+```
+
+The capabilities property should be provided using the `capabilities()` function
+imported from the same module as `setHelperManager`:
+
+```js
+import { capabilities } from '@ember/helper';
+
+class MyHelperManager {
+  capabilities = capabilities();
+}
+```
+
+Below is a description of each of the methods on the interface and their
+functions.
+
+> The remaining API docs should be copied from the descriptions in the Detailed
+> Design section of this RFC.
+
+#### `capabilities`
+
+`capabilities` returns a capabilities configuration which can be used to modify
+the behavior of the manager. Manager capabilities _must_ be provided using the
+`capabilities` function, as the underlying implementation can change over time.
+
+The first argument to capabilities is a version string, which is the version of
+Ember that the capabilities were defined in. Ember can add new versions at any
+time, and these may have entirely different behaviors, but it will not remove
+old versions until the next major version.
+
+```js
+capabilities('3.x');
+```
+
+The second argument is an optional object of capabilities and boolean values
+indicating whether they are enabled or disabled.
+
+```js
+capabilities('3.x', {
+  hasValue: true,
+  hasDestructor: true,
+});
+```
+
+If no value is specified, then the default value will be used.
+
+##### `3.x` capabilities
+
+> The remaining API docs should be copied from the descriptions in the Detailed
+> Design section of this RFC. 3.x above should be replaced with the version that
+> helper managers are initially released in.
+
+## Drawbacks
+
+- Adds a layer of indirection to helpers, which could add to complexity and
+  cost in terms of performance. This isn't likely, as we haven't seen this
+  happen with other managers we've introduced.
+
+## Alternatives
+
+- We could continue using the current helper APIs, and try to incrementally
+  migrate them to only use native classes. This wouldn't match the strategy
+  we've taken with other template constructs, like components and modifiers, and
+  would result in less ability for the community to experiment and less
+  flexibility if we chose to change helpers again in the future.
+
+- The `isScheduledEffect` capability could be broken out into a separate RFC.
+  It is mostly separable, except for the impact it has on `getValue`. Value-less
+  and effect-less helpers don't really make sense, so in isolation `getValue`
+  would probably not be an optional hook, and the `hasValue` capability wouldn't
+  exist.
+
+  Capabilities can change from version to version, so this is still not a major
+  issue, but it seems like it would be easier to add from the get go.
+
+## Appendix
```

Comment:
I mentioned this above also, would you mind adding a proposed implementation of the `{{title}}` helper from `ember-page-title` as an example of an effecting helper?